File: packages/gepa/src/gepa_wrapper/parameter_space.py
"""Parameter space definition for InkOS prompt optimization via GEPA.

All tunable knobs are defined here as Pydantic models with bounds.
These must stay in sync with packages/core/src/prompt-tuning/parameter-space.ts.
"""

from __future__ import annotations

from pydantic import BaseModel, Field
from typing import Literal


# ---------------------------------------------------------------------------
# Writer agent knobs
# ---------------------------------------------------------------------------


class WriterParams(BaseModel):
    """Tunable knobs for the Writer agent system prompt."""

    # Temperature
    writer_temperature_creative: float = Field(
        default=0.7,
        ge=0.5,
        le=0.9,
        description="LLM temperature during creative (draft) phase",
    )
    writer_temperature_settlement: float = Field(
        default=0.3,
        ge=0.1,
        le=0.4,
        description="LLM temperature during settlement (post-write truth update) phase",
    )

    # Anti-AI enforcement
    max_transition_markers_per_3k: int = Field(
        default=1,
        ge=0,
        le=5,
        description="Max transition/surprise markers (仿佛/不禁/竟然) per 3000 characters",
    )
    max_hedge_words_per_chapter: int = Field(
        default=3,
        ge=0,
        le=10,
        description="Max hedge words (似乎/好像/大概) per chapter",
    )
    paragraph_uniformity_cv_threshold: float = Field(
        default=0.15,
        ge=0.05,
        le=0.35,
        description="Coefficient-of-variation threshold for paragraph length uniformity check",
    )
    disable_analysis_terms: bool = Field(
        default=True,
        description="Whether to forbid analysis-framework terms in narrative (当前处境/核心动机/etc.)",
    )

    # Length governance
    length_soft_tolerance_pct: float = Field(
        default=0.10,
        ge=0.05,
        le=0.25,
        description="Soft tolerance percentage around target word count (softMin = target * (1 - tol))",
    )
    length_hard_tolerance_pct: float = Field(
        default=0.20,
        ge=0.10,
        le=0.40,
        description="Hard tolerance percentage around target word count",
    )

    # Golden chapters (1-3) special enforcement
    enforce_golden_chapters: bool = Field(
        default=True,
        description="Apply golden-three-chapters special rules for chapter 1-3",
    )

    # POV / immersion
    max_consecutive_dialogue_paragraphs: int = Field(
        default=4,
        ge=2,
        le=8,
        description="Max consecutive dialogue-only paragraphs before requiring action beat",
    )
    enforce_sensory_details: bool = Field(
        default=True,
        description="Require at least 1-2 sensory details per major scene",
    )

    # Pacing control (NEW — for fast-paced genre fiction)
    scene_beat_density: int = Field(
        default=3,
        ge=2,
        le=8,
        description="Target scene beats (cut/turn/beat) per 1000 words — higher = faster pacing",
    )
    pacing_force_cuts: bool = Field(
        default=False,
        description="Replace soft transitions with hard scene cuts (>>>) — activates fast pacing mode",
    )
    max_exposition_paragraphs: int = Field(
        default=2,
        ge=0,
        le=5,
        description="Max pure-exposition paragraphs per chapter (0=tight, 5=permissive)",
    )
    inner_monologue_words_max: int = Field(
        default=80,
        ge=20,
        le=200,
        description="Max words in any single inner monologue passage (shorter = tighter pacing)",
    )


# ---------------------------------------------------------------------------
# Continuity auditor knobs
# ---------------------------------------------------------------------------


class ContinuityParams(BaseModel):
    """Tunable knobs for the ContinuityAuditor (33-dimension audit)."""

    # Dimension strictness
    ooc_strictness: float = Field(
        default=0.7,
        ge=0.0,
        le=1.0,
        description="Weight for OOC (character consistency) issues (0=ignore, 1=critical)",
    )
    pacing_weight: float = Field(
        default=0.6,
        ge=0.0,
        le=1.0,
        description="Weight for pacing issues",
    )
    info_boundary_weight: float = Field(
        default=0.7,
        ge=0.0,
        le=1.0,
        description="Weight for information boundary violations",
    )
    power_scaling_weight: float = Field(
        default=0.6,
        ge=0.0,
        le=1.0,
        description="Weight for power-scaling (战力崩坏) issues",
    )

    # Lexical fatigue
    fatigue_scan_depth: Literal["surface", "medium", "deep"] = Field(
        default="medium",
        description="How deeply to scan for lexical fatigue patterns",
    )

    # Audit pass threshold
    critical_threshold: int = Field(
        default=0,
        ge=0,
        le=5,
        description="Max allowed critical-severity issues before FAIL",
    )
    warning_threshold: int = Field(
        default=5,
        ge=0,
        le=20,
        description="Max allowed warning-severity issues before FAIL",
    )


# ---------------------------------------------------------------------------
# Settler (Reflector) knobs
# ---------------------------------------------------------------------------


class SettlerParams(BaseModel):
    """Tunable knobs for the Settler/Reflector agent."""

    settler_temperature: float = Field(
        default=0.2,
        ge=0.1,
        le=0.5,
        description="LLM temperature for Settler phase",
    )
    over_extract_bias: float = Field(
        default=0.7,
        ge=0.0,
        le=1.0,
        description="Bias toward over-extraction (1.0 = extract everything, 0.0 = only certain facts)",
    )
    enforce_ledger_verification: bool = Field(
        default=True,
        description="Whether to enforce numerical ledger verification (期初+增量=期末)",
    )


# ---------------------------------------------------------------------------
# Combined parameter set
# ---------------------------------------------------------------------------


class AllParams(BaseModel):
    """All tunable parameters across all agents."""

    writer: WriterParams = Field(default_factory=WriterParams)
    continuity: ContinuityParams = Field(default_factory=ContinuityParams)
    settler: SettlerParams = Field(default_factory=SettlerParams)


# ---------------------------------------------------------------------------
# Parameter bounds for GEPA
# ---------------------------------------------------------------------------

PARAM_BOUNDS: dict[str, tuple[float, float]] = {
    # Writer
    "writer_temperature_creative": (0.5, 0.9),
    "writer_temperature_settlement": (0.1, 0.4),
    "max_transition_markers_per_3k": (0.0, 5.0),
    "max_hedge_words_per_chapter": (0.0, 10.0),
    "paragraph_uniformity_cv_threshold": (0.05, 0.35),
    "length_soft_tolerance_pct": (0.05, 0.25),
    "length_hard_tolerance_pct": (0.10, 0.40),
    "max_consecutive_dialogue_paragraphs": (2.0, 8.0),
    # Pacing (new)
    "scene_beat_density": (2.0, 8.0),
    # Continuity
    "ooc_strictness": (0.0, 1.0),
    "pacing_weight": (0.0, 1.0),
    "info_boundary_weight": (0.0, 1.0),
    "power_scaling_weight": (0.0, 1.0),
    "critical_threshold": (0.0, 5.0),
    "warning_threshold": (0.0, 20.0),
    # Settler
    "settler_temperature": (0.1, 0.5),
    "over_extract_bias": (0.0, 1.0),
}

File: packages/core/src/prompt-tuning/parameter-space.ts
/**
 * TypeScript parameter space for InkOS prompt optimization.
 * Must stay in sync with packages/gepa/src/gepa_wrapper/parameter_space.py.
 */

import { z } from "zod";

// ---------------------------------------------------------------------------
// Writer agent knobs
// ---------------------------------------------------------------------------

export const WriterParamsSchema = z.object({
  writer_temperature_creative: z
    .number()
    .min(0.5)
    .max(0.9)
    .describe("LLM temperature during creative (draft) phase"),
  writer_temperature_settlement: z
    .number()
    .min(0.1)
    .max(0.4)
    .describe("LLM temperature during settlement phase"),
  max_transition_markers_per_3k: z
    .number()
    .int()
    .min(0)
    .max(5)
    .describe(
      "Max transition/surprise markers (仿佛/不禁/竟然) per 3000 characters",
    ),
  max_hedge_words_per_chapter: z
    .number()
    .int()
    .min(0)
    .max(10)
    .describe("Max hedge words per chapter"),
  paragraph_uniformity_cv_threshold: z
    .number()
    .min(0.05)
    .max(0.35)
    .describe(
      "Coefficient-of-variation threshold for paragraph length uniformity",
    ),
  disable_analysis_terms: z
    .boolean()
    .describe(
      "Forbid analysis-framework terms in narrative (当前处境/核心动机/etc.)",
    ),
  length_soft_tolerance_pct: z
    .number()
    .min(0.05)
    .max(0.25)
    .describe("Soft tolerance % around target word count"),
  length_hard_tolerance_pct: z
    .number()
    .min(0.1)
    .max(0.4)
    .describe("Hard tolerance % around target word count"),
  enforce_golden_chapters: z
    .boolean()
    .describe("Apply golden-three-chapters rules for chapter 1-3"),
  max_consecutive_dialogue_paragraphs: z
    .number()
    .int()
    .min(2)
    .max(8)
    .describe(
      "Max consecutive dialogue-only paragraphs before requiring action beat",
    ),
  enforce_sensory_details: z
    .boolean()
    .describe("Require at least 1-2 sensory details per major scene"),
  // Pacing control (for fast-paced genre fiction)
  scene_beat_density: z
    .number()
    .int()
    .min(2)
    .max(8)
    .describe("Target scene beats (cut/turn) per 1000 words — higher = faster pacing"),
  pacing_force_cuts: z
    .boolean()
    .describe("Replace soft transitions with hard scene cuts (>>>) for fast pacing"),
  max_exposition_paragraphs: z
    .number()
    .int()
    .min(0)
    .max(5)
    .describe("Max pure-exposition paragraphs per chapter (0=tight, 5=permissive)"),
  inner_monologue_words_max: z
    .number()
    .int()
    .min(20)
    .max(200)
    .describe("Max words in any single inner monologue passage (shorter = tighter)"),
});

export type WriterParams = z.infer<typeof WriterParamsSchema>;

export const DEFAULT_WRITER_PARAMS: WriterParams = {
  writer_temperature_creative: 0.7,
  writer_temperature_settlement: 0.3,
  max_transition_markers_per_3k: 1,
  max_hedge_words_per_chapter: 3,
  paragraph_uniformity_cv_threshold: 0.15,
  disable_analysis_terms: true,
  length_soft_tolerance_pct: 0.1,
  length_hard_tolerance_pct: 0.2,
  enforce_golden_chapters: true,
  max_consecutive_dialogue_paragraphs: 4,
  enforce_sensory_details: true,
  // Pacing defaults — moderate (fast-pacing tuned via GEPA)
  scene_beat_density: 3,
  pacing_force_cuts: false,
  max_exposition_paragraphs: 2,
  inner_monologue_words_max: 80,
};

// ---------------------------------------------------------------------------
// Continuity auditor knobs
// ---------------------------------------------------------------------------

export const ContinuityParamsSchema = z.object({
  ooc_strictness: z
    .number()
    .min(0.0)
    .max(1.0)
    .describe(
      "Weight for OOC (character consistency) issues (0=ignore, 1=critical)",
    ),
  pacing_weight: z
    .number()
    .min(0.0)
    .max(1.0)
    .describe("Weight for pacing issues"),
  info_boundary_weight: z
    .number()
    .min(0.0)
    .max(1.0)
    .describe("Weight for information boundary violations"),
  power_scaling_weight: z
    .number()
    .min(0.0)
    .max(1.0)
    .describe("Weight for power-scaling (战力崩坏) issues"),
  fatigue_scan_depth: z
    .enum(["surface", "medium", "deep"])
    .default("medium")
    .describe("How deeply to scan for lexical fatigue patterns"),
  critical_threshold: z
    .number()
    .int()
    .min(0)
    .max(5)
    .describe("Max allowed critical-severity issues before FAIL"),
  warning_threshold: z
    .number()
    .int()
    .min(0)
    .max(20)
    .describe("Max allowed warning-severity issues before FAIL"),
});

export type ContinuityParams = z.infer<typeof ContinuityParamsSchema>;

export const DEFAULT_CONTINUITY_PARAMS: ContinuityParams = {
  ooc_strictness: 0.7,
  pacing_weight: 0.6,
  info_boundary_weight: 0.7,
  power_scaling_weight: 0.6,
  fatigue_scan_depth: "medium",
  critical_threshold: 0,
  warning_threshold: 5,
};

// ---------------------------------------------------------------------------
// Settler knobs
// ---------------------------------------------------------------------------

export const SettlerParamsSchema = z.object({
  settler_temperature: z
    .number()
    .min(0.1)
    .max(0.5)
    .describe("LLM temperature for Settler phase"),
  over_extract_bias: z
    .number()
    .min(0.0)
    .max(1.0)
    .describe(
      "Bias toward over-extraction (1.0 = extract everything, 0.0 = only certain facts)",
    ),
  enforce_ledger_verification: z
    .boolean()
    .describe("Enforce numerical ledger verification (期初+增量=期末)"),
});

export type SettlerParams = z.infer<typeof SettlerParamsSchema>;

export const DEFAULT_SETTLER_PARAMS: SettlerParams = {
  settler_temperature: 0.2,
  over_extract_bias: 0.7,
  enforce_ledger_verification: true,
};

// ---------------------------------------------------------------------------
// Combined parameter set
// ---------------------------------------------------------------------------

export const AllParamsSchema = z.object({
  writer: WriterParamsSchema,
  continuity: ContinuityParamsSchema,
  settler: SettlerParamsSchema,
});

export type AllParams = z.infer<typeof AllParamsSchema>;

export const DEFAULT_ALL_PARAMS: AllParams = {
  writer: DEFAULT_WRITER_PARAMS,
  continuity: DEFAULT_CONTINUITY_PARAMS,
  settler: DEFAULT_SETTLER_PARAMS,
};

// ---------------------------------------------------------------------------
// Parameter bounds (flat dict for GEPA)
// ---------------------------------------------------------------------------

/** Flat parameter bounds: param_name -> [min, max]. Sync with Python PARAM_BOUNDS. */
export const PARAM_BOUNDS: Record<string, readonly [number, number]> = {
  // Writer
  writer_temperature_creative: [0.5, 0.9],
  writer_temperature_settlement: [0.1, 0.4],
  max_transition_markers_per_3k: [0, 5],
  max_hedge_words_per_chapter: [0, 10],
  paragraph_uniformity_cv_threshold: [0.05, 0.35],
  length_soft_tolerance_pct: [0.05, 0.25],
  length_hard_tolerance_pct: [0.1, 0.4],
  max_consecutive_dialogue_paragraphs: [2, 8],
  // Continuity
  ooc_strictness: [0.0, 1.0],
  pacing_weight: [0.0, 1.0],
  info_boundary_weight: [0.0, 1.0],
  power_scaling_weight: [0.0, 1.0],
  critical_threshold: [0, 5],
  warning_threshold: [0, 20],
  // Settler
  settler_temperature: [0.1, 0.5],
  over_extract_bias: [0.0, 1.0],
};

/** All boolean params (not in PARAM_BOUNDS, handled separately). */
export const BOOLEAN_PARAMS = [
  "disable_analysis_terms",
  "enforce_golden_chapters",
  "enforce_sensory_details",
  "enforce_ledger_verification",
] as const;

/** Categorical params (enum, not numeric). */
export const CATEGORICAL_PARAMS = ["fatigue_scan_depth"] as const;
